dlib/metaprogramming.h

```diff
@@ -3,46 +3,28 @@
 #ifndef DLIB_METApROGRAMMING_Hh_
 #define DLIB_METApROGRAMMING_Hh_
 
-#include "algs.h"
+#include "constexpr_if.h"
+#include "invoke.h"
 
 namespace dlib
 {
+// ----------------------------------------------------------------------------------------
+    namespace impl
+    {
+        template<typename List>
+        struct add_one {};
+
+        template<std::size_t... n>
+        struct add_one<std::index_sequence<n...>> { using type = std::index_sequence<(n+1)...>; };
+
+        template<typename List>
+        using add_one_t = typename add_one<List>::type;
+    }
 
 // ----------------------------------------------------------------------------------------
 
     template <size_t... n>
-    struct compile_time_integer_list 
-    {
-        /*!
-            WHAT THIS OBJECT REPRESENTS
-                The point of this type is to, as the name suggests, hold a compile time list of integers.
-                As an example, here is something simple you could do with it:
-
-                    template <size_t... ints>
-                    void print_compile_time_ints (
-                        compile_time_integer_list<ints...>
-                    )
-                    {
-                        print(ints...);
-                    }
-
-                    int main()
-                    {
-                        print_compile_time_ints(compile_time_integer_list<0,4,9>());
-                    }
-
-                Which just calls: print(0,4,9);
-
-                This is a simple example, but this kind of thing is useful in larger and
-                more complex template metaprogramming constructs.
-        !*/
-
-        template <size_t m>
-        struct push_back
-        {
-            typedef compile_time_integer_list<n..., m> type;
-        };
-    };
+    using compile_time_integer_list = std::index_sequence<n...>;
```

Comment:
This isn't the same thing since it misses `push_back`.  So this would break users using that.

Reply:
Yeah I was wondering if push_back was there only for implementation details. I was thinking of putting a deprecation notice on that and all the dlib type traits.

Reply:
Na don't say it's deprecated.  Deprecating stuff just upsets the oldest and most loyal users.  Like if something needed to change due to some specific reason then sure.  Like if CUDA was dropping a feature we would necessarily also need to react to that and that might force some API change.  But this class existing isn't bothering anything.

Reply:
Ok fine. "Oldest and most loyal users"? :) Sounds like dlib is specifically built for 1 or 2 key users :) Also, shouldn't we encourage people to update their code and use best practices ? Issuing a compiler warning isn't a big deal. If they build their code with -Werror they will realise that it was about time they updated their code, even more so since updating their compiler to something that supports c++14. Also I'm pretty sure they won't be too upset. Anyway, I don't I'm not too opinionated about deprecation warnings. I don't use them myself but I sure do appreciate it when I see them. They're often educational too.

Reply:
I'll put push_back in. I'm away for a week so it will have to wait. Sorry

Reply:
Ha.  Na, it's that there are a lot of people using it in products and otherwise at work.  It sucks to have to spend time (i.e. incur a cost to your business) because someone else is asking you to update code that is working perfectly fine for your needs.